### PR TITLE
Updates to storage proposal with more storage contract details

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Replicator proposal should have all the detail of how the storage contract works.

#### Summary of Changes

List all transaction types relevant to the contract(s) and what checks/state transitions should be performed.

Fixes #

### CODE REVIEW

**Comment by sakridge** [comment]
@sagar-solana is it what you are thinking of?

File: proposals/src/ledger-replication-to-implement.md [hunk 1]
```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+    signature: Signature,
+};
+keys = [replicator_keypair]
+```
+Replicators create these after mining their stored ledger data for a certain hash value.
+The slot is the end slot of the segment of ledger they are storing, the sha\_state
+the result of the replicator using the hash function to sample their encrypted ledger segment.
+The signature is the signature that was created when they signed a PoH value for the
+current storage epoch. The list of proofs from the current storage epoch should be saved
+in the account state, and then transfered to a list of proofs for the previous epcoh when
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: typo: epcoh

File: proposals/src/ledger-replication-to-implement.md [hunk 2]
```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+    sha_state: Hash,
+    signature: Signature,
+};
+keys = [replicator_keypair]
+```
+Replicators create these after mining their stored ledger data for a certain hash value.
+The slot is the end slot of the segment of ledger they are storing, the sha\_state
+the result of the replicator using the hash function to sample their encrypted ledger segment.
+The signature is the signature that was created when they signed a PoH value for the
+current storage epoch. The list of proofs from the current storage epoch should be saved
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: What defines how long a storage epoch lasts ? compared to a regular cluster epoch?

File: proposals/src/ledger-replication-to-implement.md [hunk 3]
```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+in the account state, and then transfered to a list of proofs for the previous epcoh when
+the epoch passes. In a given storage epoch a given replicator should only submit proofs
+for one segment. The program should check that the slot and signature is valid and then store
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: The slot is valid as long as it belongs to the current storage epoch? How is the signature verified, the program will need to know what storage_blockhash was used. Is this related to the storage_epoch we're in ? sorry I know some of these might be really basic questions

File: proposals/src/ledger-replication-to-implement.md [hunk 4]
```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+a record of this proof being submitted in the current storage epoch.
+
+#### ProofValidation
+```compile_fail
+ProofValidation {
+   proof_mask: Vec<ProofStatus>,
+}
+keys = [validator_keypair, replicator_keypair(s) (unsigned)]
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: We also need to carry information about which proofs map to which replicator. I don't see how to do that here without some implicit ordering of account keys and proof_mask. Thats the only way to do it without adding more information in the proof_mask like we discussed earlier. That's the intended design, yes?

Comment (sagar-solana) [reply proposals/src/ledger-replication-to-implement.md]: Also some information about which segment this is for. Otherwise we'd have to pass in a proof_mask vec the size of the storage epoch where everything outside the current segment is `Skip`. Instead if we pass in the segment for which this validation was performed, the mask can be made tiny.

File: proposals/src/ledger-replication-to-implement.md [hunk 5]
```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+```compile_fail
+ClaimStorageReward {
+}
+keys = [validator_keypair or replicator_keypair, validator/replicator_keypairs (unsigned)]
+```
+Replicators and validators will use this transaction to get paid tokens from a program state
+where SubmitStorageProof, ProofValidation and ChallengeProofValidations are in a state where
+proofs have been submitted and validated and there are no ChallengeProofValidations referencing
+those proofs. For a validator, it should reference the replicator keypairs to which it has validated
+proofs in the relevant epoch. And for a replicator it should reference validator keypairs for which it
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: replicators are totally unaware of which validators submitted proofs for them. how should they keep track of this? They don't even know what keypair the validator is using since that keypair will be hidden and not related the validator's id which is available via gossip.

File: proposals/src/ledger-replication-to-implement.md [hunk 6]
```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+has validated and wants to be rewarded.
+
+#### ChallengeProofValidation
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: oh I dont think I understand this at all. Why do replicators submit fake proofs? shouldn't that be slash-able?

Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: They do it to prevent validators from just approving everything they see. The value they use for the hash though is a short value made up by the replicator, then the replicator can show the network the seed value it used to create the hash to catch the validator. In the case it catches the validator, the validator should not be paid for any ProofValidations that it produces in that epoch.

File: proposals/src/ledger-replication-to-implement.md (hunk 5 above)
Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: I think they can ask some random validator(s) for the set of ProofValidations from the previous epoch at some point for the segment that they submitted storage proofs for.

File: proposals/src/ledger-replication-to-implement.md (hunk 6 above)
Comment (sagar-solana) [reply proposals/src/ledger-replication-to-implement.md]: Ah I see. So with every set of proofs in an epoch, some of them must be fake so that replicators keep validators in check. Validators need to be sure to keep sampling replicators every now and then too.

Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: Yes, that's right.

File: proposals/src/ledger-replication-to-implement.md (hunk 4 above)
Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: The referenced replicator_keypair in `accounts[1]` will give them the set of storage proofs. Since a replicator keypair can only submit proofs for a single segment in a storage epoch time, then it specifies which segment it is for in that way. Could make it explicit, but I don't think it is strictly needed.

File: proposals/src/ledger-replication-to-implement.md (hunk 2 above)
Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: TBD, but thinking longer than the cluster epoch, around the time (or number of slots) it takes to generate somewhere around 100GB to 1TB of ledger. At full speed, 1TB is ~2 hours, 100GB is ~15 minutes.

File: proposals/src/ledger-replication-to-implement.md (hunk 3 above)
Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: I added a line about that, we can keep a set of valid slots.

**Comment by sakridge** [comment]
@sagar-solana take another pass when you can. Thanks.

**Comment by sagar-solana** [comment]
I will, sorry spent all day trying to work around the packet drops we're seeing. Seems like a kernel bug we're dealing with...

File: book/src/ledger-replication.md [hunk 7]
```diff
@@ -31,7 +31,7 @@ core. The total space required for verification is `1_ledger_segment +
 Validators for PoRep are the same validators that are verifying transactions.
 They have some stake that they have put up as collateral that ensures that
 their work is honest. If you can prove that a validator verified a fake PoRep,
-then the validators stake can be slashed.
+then the validator will not receive a reward for that storage epoch.
```
Comment (sagar-solana) [inline book/src/ledger-replication.md]: @garious, I'm a little confused about how this ties in with the vote accounts. A validator's stake usually lives in the vote account. Will the storage rewards need to be aware of the reward accounts for all delegated stakes? Just trying to understand how the two rewards fit together and whether they even need to.

Comment (garious) [reply book/src/ledger-replication.md]: The stake only currently lives in the vote account. The Passive Delegated Stakes proposal moves them to a separate account so that the validator can send votes to just one account, no matter how many stakers delegate their stake to that validator.

Comment (sagar-solana) [reply book/src/ledger-replication.md]: Okay so that should help simplify this a little. Maybe the storage account can carry the id that passive delegated account as well instead of us having to loop through all accounts associated with a node (whether via delegation or direct ownership). 

Our current plan is to setup a separate storage account for validators. Using a new keypair. In the current proposal these accounts don't carry any information about who their owner (validator) is. I think we'll need to add that though so we can correctly tie things together.  @sakridge thoughts ?

File: proposals/src/ledger-replication-to-implement.md [hunk 8]
```diff
@@ -37,3 +43,100 @@ transacation proves the validator incorrectly validated a fake storage proof.
 The replicator is rewarded and the validator's staking balance is slashed or
 frozen.
 
+### Storage proof contract logic
+
+Each replicator and validator will have their own storage account. The validator's
+account would be separate from their gossip id similiar to their vote account.
```
Comment (sagar-solana) [inline proposals/src/ledger-replication-to-implement.md]: This has some maintenance problems, we need to store this keypair somewhere :( On the first pass I can just `keypair::new()` everytime, but aside from that they'll need to pass in a keypair to the validator somehow (like with vote accounts - but that thing is another monster). No way we can just reuse the validator's primary keypair right ?

File: proposals/src/ledger-replication-to-implement.md [hunk 9]
```diff
@@ -37,3 +43,100 @@ transacation proves the validator incorrectly validated a fake storage proof.
...
+These should be implemented as two programs one which handles the validator as the keysigner
+and one for the replicator. In that way when the programs reference other accounts, they
+can check the program id to ensure it is a validator or replicator account they are
+referencing.
+
+#### SubmitMiningProof
+```compile_fail
```
Comment (garious) [inline proposals/src/ledger-replication-to-implement.md]: `rust,ignore` is what you're looking for there

File: book/src/ledger-replication.md (hunk 7 above)
Comment (sakridge) [reply book/src/ledger-replication.md]: What is the scenario where they need to?

File: proposals/src/ledger-replication-to-implement.md (hunk 8 above)
Comment (sakridge) [reply proposals/src/ledger-replication-to-implement.md]: The primary keypair would have to be an account where account.owner == storage_program, so  I'm not sure if that's possible or *should* be the case for the validator's primary keypair.

File: book/src/ledger-replication.md (hunk 7 above)
Comment (sakridge) [reply book/src/ledger-replication.md]: Current design is to not need slashing.

Comment (sagar-solana) [reply book/src/ledger-replication.md]: For example, we'd need to co-related validator <-> storage_account to figure out how to deliver rewards

Comment (sakridge) [reply book/src/ledger-replication.md]: The validator has the keypair for the storage account, it can use that to transfer the rewards anywhere it wants.
